feat(EventTracker): Use sendBeacon instead of Fetch with dispatch on unload rather than on next load.

klarstrup opened the pull request:
Closes #161

klarstrup committed:
feat(EventTracker): Replace persisted pool with sendBeacon

klarstrup commented on lib/kits/events/tracker.ts:
```diff
@@ -239,14 +242,14 @@ class Tracker {
         if (this.eventsTrackUrl) {
-            dispatch(this.eventsTrackUrl, this.eventsTrackHeaders);
+            this.dispatch();
             this.hasMadeInitialDispatch = true;
```
This on-instantiation dispatching and tracking of `hasMadeInitialDispatch` we can also get rid of if there aren't events(from a persisted pool) to dispatch immediately when `new Tracker`, but we would want to cover that for a transition for any backed-up event pool for a week or whatever

klarstrup committed:
refactor(EventTracker): Only use sendBeacon, ditch validation and retrying

mortenbo commented:
@klarstrup I think it looks really good. Let's start out by testing this in Vizsla staging before we release this anywhere else.

klarstrup commented:
Then I'd like to see you and @jrlarano merge your approved PRs so those can be released and this can be moved ahead into beta for Vizsla's usage

mortenbo commented:
I've deleted my PR as it's not relevant after all 👍

klarstrup committed:
chore(browserslist): require supports beacon

klarstrup committed:
Merge branch 'develop' into io/sendbeacon

klarstrup committed:
Merge branch 'develop' into io/sendbeacon

klarstrup committed:
feat(Tracker): Add best-effort sendBeacon ponyfill

This regains us some coverage without very much overhead.

Ita also allows running the EventsKit tests in our sendBeacon-less node(fetch)/jsdom(XMLHttpRequest) test environments.

klarstrup committed:
fix(Tracker): Don't check visibilityState on beforeUnload

klarstrup committed:
Merge branch 'develop' into io/sendbeacon

klarstrup commented on lib/kits/events/tracker.ts:
```diff
@@ -351,70 +394,16 @@ class Tracker {
-            return;
+    dispatching = false;
+    dispatchLimit = 100;
+    dispatch = throttle(this.dispatchBeacon, 4000);
```
Going to swap this `throttle` for a more suitable `debounce`, such that we don't do a lot of immediate single-event dispatches.

klarstrup committed:
Merge branch 'develop' into io/sendbeacon

klarstrup committed:
tests(EventsKit): Test onBeforeUnload dispatch

Sharpen the delayed dispatch tests

klarstrup committed:
feat(Tracker): Defer dispatch with debounce

Throttle results in a single event instantly dispatching without batching

Augmented throttle types too

klarstrup committed:
Merge branch 'develop' into io/sendbeacon

klarstrup committed:
test(events): Be more patient with beforeUnload dispatch

klarstrup committed:
test(events): Be more patient with beforeUnload dispatch

klarstrup committed:
test(events): Be more patient with beforeUnload dispatch

klarstrup committed:
test(events): Be more patient with beforeUnload dispatch

klarstrup committed:
test(events): goto blank instead of close page
